### PR TITLE
Updated README.md

### PR DESCRIPTION
**Pull request by Mayank447**
Updated README.md in sync with derivative 1D class

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,9 +1,159 @@
...
+
+- After setting up the package, you can import the classes and methods from `pseudospectral` module in your code as follows:
+  ``from pseudospectral import DiracOperator, FreeFermions2D``
+- You may further refer to:
+  - `example_scripts` folder for example script usage of the DiracOperator, Spectrum classes, methods and a boiler-plate to begin with.
+  - `tutorial.md` for a step-by-step guide on how to use the classes and methods and a bit of theory behind the implementation (history behind the project).
+
+## DiracOperator Class
+
+The DiracOperator class provides several methods to work with the Dirac operator in both real and spectral domains.
```
Comment (chillenzer) [inline README.md:50]: That is not a helpful text. This information is obvious from the name. Better provide some useful information. What does "work with" mean? Maybe something like this
```suggestion
The `DiracOperator` class provides wrapping infrastructure to apply an operator defined by its `Spectrum` to a function defined by its vector of `coefficients` in `"real"` or `"spectral"` basis.
```

File: README.md [hunk 2]
```diff
@@ -1,9 +1,159 @@
...
+
+#### 1.1 Args:
+
+- spectrum: The spectrum class object that represents the spectrum of the given system / operator.
+
+### 2 Methods
+
+#### 2.1 ``apply_to``
+
+Applies the given Dirac operator to the input coefficients as per the given input and output spaces.
```
Comment (chillenzer) [inline README.md:62]:
```suggestion
Applies the given Dirac operator to a function represented by its coefficient vector in the specified `"real"` or `"spectral"` basis.
```

File: README.md [hunk 3]
```diff
@@ -1,9 +1,159 @@
...
+
+##### 2.1.1 Signature: ``apply_to(coefficients, input_basis='real', output_basis='real')``
+
+##### 2.1.2 Parameters:
+
+- coefficients (np.ndarray): An array of coefficients in the input space.
```
Comment (chillenzer) [inline README.md:68]:
```suggestion
- coefficients (np.ndarray): An array of coefficients in the input basis.
```

File: README.md [hunk 4]
```diff
@@ -1,9 +1,159 @@
...
+- input_basis (str): Specifies the space of the input coefficients. Valid values are 'real' and 'spectral'.
+- output_basis (str): Specifies the space of the output coefficients. Valid values are 'real' and 'spectral'.
+
+##### 2.1.3 Returns:
+
+- np.ndarray: An array of coefficients in the output space.
```
Comment (chillenzer) [inline README.md:74]:
```suggestion
- np.ndarray: An array of coefficients in the output basis.
```
Everywhere

File: README.md [hunk 5]
```diff
@@ -1,9 +1,159 @@
...
+
+##### 2.1.4 Raises:
+
+- ValueError: If input_basis or output_basis is not 'real' or 'spectral'.
+
+## Derivative1D Class
+
+The Derivative1D class provides methods to work with the 1D derivative operator in both real and spectral domains with periodic boundary conditions.
```
Comment (chillenzer) [inline README.md:82]: As above.

File: README.md [hunk 6]
```diff
@@ -1,9 +1,159 @@
...
+
+#### 3.1 ``eigenfunctions``
+
+Returns the eigenfunction of the 1D derivative operator at the given eigenvalue index i.e. exp(ikx).
+
+##### 3.1.1 Signature: `eigenfunctions(eigenvalue_index)`
+
+##### 3.1.2 Parameters:
+
+- eigenvalue_index (int): The index of the eigenvalue. This range is from 0 to L - 1.
```
Comment (chillenzer) [inline README.md:107]:
```suggestion
- eigenvalue_index (int): The index of the eigenvalue. This range is from 0 to `num_lattice_points - 1`.
```
Everywhere

File: README.md [hunk 7]
```diff
@@ -1,9 +1,159 @@
...
+
+- lambda function: The eigenfunction of the 1D derivative operator which can be sampled at the given points by passing the array as an argument.
+
+##### 3.1.4 Raises:
+
+- ValueError: If eigenvalue_index is not in the range from 0 to L - 1.
+
+#### 3.2 ``_eigenvalues``
+
+Private function to compute the eigenvalues of the 1D derivative operator.
```
Comment (chillenzer) [inline README.md:119]: Private functions are private and no public interface. Don't advertise them in a public section.

File: README.md [hunk 8]
```diff
@@ -1,9 +1,159 @@
...
+
+- np.ndarray: An array of coefficients in the output space.
+
+##### 3.3.4 Raises:
+
+- ValueError: If input_basis or output_basis is not 'real' or 'spectral'.
+
+#### 3.4 ``lattice``
+
+Generates a discretized set of points representing the lattice structure in either the time or frequency domain.
```
Comment (chillenzer) [inline README.md:143]:
```suggestion
Generates a discretized set of points representing the lattice structure in either the `"real"` or `"spectral"` domain. This defines the basis with respect to which the coefficients are to be interpreted, e.g., the `i`th eigenfunction in real-space basis representation is given by `spectrum.eigenfunction(i)(spectrum.lattice("real")).`
```

**Review by chillenzer — CHANGES_REQUESTED** [review]
Thanks for the effort.

Please keep in mind two things:
- Whenever you add explanation, comments, etc., to your project it should add more value than noise. A sentence vaguely saying "Something provides methods to work with <what's already in the name of Something>." is simply noise. In the best case, the user skips this sentence because it is obvious. In the worst case, it gets out of sync with the actual code and misleads the reader.
- Please have a look at the amount of mistakes and out-of-sync-ness that's already in this first version of the `README` and re-consider if you really want to maintain this level of detail twice (because we also want to have docstrings saying exactly the same).

File: README.md [hunk 9]
```diff
@@ -1,9 +1,159 @@
...
+├── example_scripts
+    ├── cosine_derivative.py
+├── LICENSE
+└── tests
+    ├── conftest.py
+    ├── test_real_basis.py
+    ├── test_spectral_basis.py
+    ├── test_transform.py
+    ├── test_spectra.py
+`</pre>`
```
[the remark this answers is not in the document]
Comment (Mayank447) [reply README.md]: Fixed

File: README.md [hunk 10]
```diff
@@ -1,9 +1,159 @@
...
+
+#### Setup
+
+- git clone the repository - `git clone https://github.com/Mayank447/Pseudo-Spectral-Discretization.git`
+- cd into the repository - `cd Pseudo-Spectral-Discretization`
+- run `pip install .`
```
[the remark this answers is not in the document]
Comment (Mayank447) [reply README.md]: Changed

File: README.md [hunk 11]
```diff
@@ -1,9 +1,159 @@
...
+
+#### Usage
+
+- After setting up the package, you can import the classes and methods from `pseudospectral` module in your code as follows:
+  ``from pseudospectral import DiracOperator, FreeFermions2D``
+- You may further refer to:
+  - `example_scripts` folder for example script usage of the DiracOperator, Spectrum classes, methods and a boiler-plate to begin with.
```
[the remark this answers is not in the document]
Comment (Mayank447) [reply README.md:44]: Changed

File: README.md [hunk 12]
```diff
@@ -1,9 +1,159 @@
...
+  - `tutorial.md` for a step-by-step guide on how to use the classes and methods and a bit of theory behind the implementation (history behind the project).
```
[the remark this answers is not in the document]
Comment (Mayank447) [reply README.md:45]: I am not done with tutorial.md

File: README.md (hunk 2 above)
Comment (Mayank447) [reply README.md:61]: Fixed

File: README.md (hunk 1 above)
Comment (Mayank447) [reply README.md:49]: Fixed

File: README.md [hunk 13]
```diff
@@ -1,9 +1,159 @@
...
+
+## DiracOperator Class
+
+The DiracOperator class provides several methods to work with the Dirac operator in both real and spectral domains.
+
+### 1 Initialization ``DiracOperator(spectrum)``
+
+#### 1.1 Args:
+
+- spectrum: The spectrum class object that represents the spectrum of the given system / operator.
```
[the remark this answers is not in the document]
Comment (Mayank447) [reply README.md:55]: Changed

File: README.md (hunk 3 above)
Comment (Mayank447) [reply README.md:67]: Changed

File: README.md (hunk 4 above)
Comment (Mayank447) [reply README.md:73]: Changed

File: README.md (hunk 7 above)
Comment (Mayank447) [reply README.md:118]: Changes made

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (chillenzer) [reply README.md:38]: Ah, sorry. Then, it's something like `./Pseudo-Spectral-Discretization` but it's not important.

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (chillenzer) [reply README.md:46]: Then remove that piece of information for now and add it once it's correct.

File: README.md [hunk 14]
```diff
@@ -1,9 +1,155 @@
 # Pseudo-Spectral-Discretization
 
-# DiracOperator Class
+This codebase contains the implementation of the DiracOperator class, which is used for applying the Dirac operator to spectral coefficients, computing eigenvalues and eigenfunctions, generating lattice structures, and transforming values between real and spectral spaces.
 
-The `DiracOperator` class implements the Dirac operator with methods for applying it to vectors in real-space and spectral representation. It takes a spectrum as defined below.
+### Directory Structure
 
-# Spectrum
+- `pseudospectral` folder contains all the source code of the project which are mainly `.py` files broadly categorized into Dirac Operator and Spectra python files.
```
Comment (chillenzer) [inline README.md:7]:
```suggestion
- `pseudospectral` folder contains all the source code of the project which are mainly `.py` files broadly categorized into Dirac operator and Spectra python files.
```

File: README.md [hunk 15]
```diff
@@ -1,9 +1,155 @@
...
+- `pseudospectral/spectra` folder contains spectrum class implementations.
+  - `example_scripts` folder for example script usage of the `DiracOperator`, `Spectrum` classes, methods and a boiler-plate to begin with.
```
Comment (chillenzer) [inline README.md:9]:
```suggestion
- `example_scripts` folder for example script usage of the `DiracOperator`, `Spectrum` classes, methods and a boiler-plate to begin with.
```

File: README.md [hunk 16]
```diff
@@ -1,9 +1,155 @@
...
+- `tests` folder contains all the pytests written for the project.
 
-Spectra are defined as classes which provide, among other things, knowledge about the eigenvalues and the transformations between real and spectral space.
\ No newline at end of file
+<pre>
+.
+├── pseudospectral
+    ├── __init__.py
+    ├── dirac_operator.py
+    ├── spectra
```
Comment (chillenzer) [inline README.md:17]:
```suggestion
    └── spectra
```

File: README.md [hunk 17]
```diff
@@ -1,9 +1,155 @@
...
+        ├── derivative_1D.py
+        ├── free_fermions_2D.py
```
Comment (chillenzer) [inline README.md:19]:
```suggestion
        └── free_fermions_2D.py
```
and so on...

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (chillenzer) [reply README.md:38]: But even better: https://pip.pypa.io/en/stable/topics/vcs-support/
Something like this should work: (Figure out the details yourself if you like or leave it as is)
```bash
python -m pip install 'SomeProject@git+https://git.repo/some_pkg.git@1.3.1'
```
(from bullet 5 [here](https://pip.pypa.io/en/stable/cli/pip_install/#examples).

File: README.md [hunk 18]
```diff
@@ -1,9 +1,155 @@
...
+- git clone the repository - `git clone https://github.com/Mayank447/Pseudo-Spectral-Discretization.git`
+- cd into the repository - `cd Pseudo-Spectral-Discretization`
+- run `pip install .`
+
+#### Usage
+
+- After setting up the package, you can import the classes and methods from `pseudospectral` module in your code as follows:
+  ``from pseudospectral import DiracOperator, FreeFermions2D``
+- You may further refer to:
+  - `example_scripts` folder for example script usage of the DiracOperator, Spectrum classes, methods and a boiler-plate to begin with.
```
Comment (chillenzer) [inline README.md:45]:
```suggestion
  - `example_scripts` folder for example script usage of the DiracOperator, Spectrum classes, methods and a boilerplate to begin with.
```

File: README.md [hunk 19]
```diff
@@ -1,9 +1,155 @@
...
+
+#### 2.1 ``apply_to``
+
+Applies the given Dirac operator to a function represented by its coefficient vector in the specified `"real"` or `"spectral"` basis.
+
+##### 2.1.1 Signature: ``apply_to(coefficients, input_basis='real', output_basis='real')``
+
+##### 2.1.2 Parameters:
+
+- coefficients (np.ndarray): An array of coefficients in the input basis.
```
Comment (chillenzer) [inline README.md:68]:
```suggestion
- `coefficients` (`np.ndarray`): An array of coefficients in the input basis.
```
"Everywhere" meant whenever you've got a code entity.

File: README.md [hunk 20]
```diff
@@ -1,9 +1,155 @@
...
+- input_basis (str): Specifies the space of the input coefficients. Valid values are 'real' and 'spectral'.
```
Comment (chillenzer) [inline README.md:69]:
```suggestion
- input_basis (str): Specifies the basis of the input coefficients. Valid values are 'real' and 'spectral'.
```
everywhere.

File: README.md [hunk 21]
```diff
@@ -1,9 +1,155 @@
...
+
+## Derivative1D Class
+
+The Derivative1D class provides methods to work with the 1D derivative operator in both real and spectral basis with periodic boundary conditions.
+
+### 1 Initialization ``Derivative1D``
+
+#### 1.1 Args:
+
+- L: The length of the domain.
```
Comment (chillenzer) [inline README.md:88]:
```suggestion
- L: The physical length of the domain.
```
everywhere.

File: README.md [hunk 22]
```diff
@@ -1,9 +1,155 @@
...
+
+- eigenvalue_index (int): The index of the eigenvalue. This range is from 0 to `num_lattice_points - 1`.
+
+##### 3.1.3 Returns:
+
+- lambda function: The eigenfunction of the 1D derivative operator which can be sampled at the given points by passing the array as an argument.
+
+##### 3.1.4 Raises:
+
+- ValueError: If eigenvalue_index is not in the range from 0 to L - 1.
```
Comment (chillenzer) [inline README.md:115]: Also here!

File: README.md (hunk 5 above)
Comment (chillenzer) [reply README.md]: This meant: Please add some useful information here as done for the `DiracOperator` case.

File: README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (chillenzer) [reply README.md]: Not everywhere. It meant: For all code entities.

File: README.md (hunk 6 above)
Comment (chillenzer) [reply README.md]: Not done everywhere!

**Review by chillenzer — CHANGES_REQUESTED** [review]
Here's another round. Please be particularly vigilant concerning the meaning of "everywhere". Don't just search-replace one expression but think about what this change means and how it is to apply "everywhere".

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (chillenzer) [reply README.md:46]: Okay. I'll give you this one.

**Review by chillenzer — APPROVED** [review]

**Pull request merged**